src/psyclone/psyir/nodes/call.py

```diff
@@ -593,3 +600,172 @@ def _location_txt(node):
+                f" than caller (routine '{routine.name}')"
+            )
+
+        assert len(self.arguments) == len(self.argument_names)
+
+        # Iterate over all arguments
+        ret_arg_idx_list = []
+        for call_arg_idx, call_arg in enumerate(self.arguments):
+            call_arg_idx: int
+            call_arg: DataSymbol
+
+            # If None, it's a positional argument => Just return the index if
+            # the types match
+            if self.argument_names[call_arg_idx] is None:
+                routine_arg = routine_argument_list[call_arg_idx]
+                routine_arg: DataSymbol
+
+                # Do the types of arguments match?
+                #
+                # TODO #759: If optional is used, it's an unsupported Fortran
+                # type and we need to use the following workaround
+                # Once this issue is resolved, simply remove this if branch
+                if not isinstance(
+                            routine_arg.datatype,
+                            UnsupportedFortranType):
+                    if call_arg.datatype != routine_arg.datatype:
+                        raise CallMatchingArgumentsNotFound(
+                            f"Argument type mismatch of call argument "
+                            f"'{call_arg}' and routine argument "
+                            f"'{routine_arg}'"
+                        )
+
+                ret_arg_idx_list.append(call_arg_idx)
+                routine_argument_list[call_arg_idx] = None
+                continue
+
+            #
+            # Next, we handle all named arguments
+            #
+            arg_name = self.argument_names[call_arg_idx]
+            named_arg_found = False
+            routine_arg_idx = None
+            for routine_arg_idx, routine_arg in enumerate(
+                                routine_argument_list):
+                routine_arg: DataSymbol
+
+                # Check if argument was already processed
+                if routine_arg is None:
+                    continue
+
+                if arg_name == routine_arg.name:
+                    # TODO #759: If optional is used, it's an unsupported
+                    # Fortran type and we need to use the following workaround
+                    # Once this issue is resolved, simply remove this if branch
+                    if not isinstance(
+                                routine_arg.datatype,
+                                UnsupportedFortranType):
+                        if call_arg.datatype != routine_arg.datatype:
+                            raise CallMatchingArgumentsNotFound(
+                                f"Argument type mismatch of call argument "
+                                f"'{call_arg}' and routine argument "
+                                f"'{routine_arg}'"
+                            )
+
+                    ret_arg_idx_list.append(routine_arg_idx)
+                    named_arg_found = True
```

Comment:
This pattern can be achieved without the bool:
```python
    for something in my_search:
        if found_it:
            break
    else:
        # Didn't find it.
```

Reply:
Nice solution!